### PR TITLE
Add etherToken.getLogsAsync and etherToken.subscribe with tests

### PR DESCRIPTION
**Pull request by LogvinovLeon**
This PR:
* Adds `zeroEx.etherToken.subscribe`
* Adds `zeroEx.etherToken.unsubscribe`
* Adds `zeroEx.etherToken.unsubscribeAll`
* Adds `zeroEx.etherToken.getLogsAsync`
* Adds tests for all above
* Updates orderWatcher with the new events
* Updates etherToken ABI
* Remove our EtherToken from contracts
* Adds makers WETH9
* Changes migrations to migrate the correct one

### CODE REVIEW

File: packages/0x.js/test/ether_token_wrapper_test.ts [hunk 1]
```diff
@@ -113,4 +129,157 @@ describe('EtherTokenWrapper', () => {
...
+            const tokenUtils = new TokenUtils(tokens);
+            const etherToken = tokenUtils.getWethTokenOrThrow();
+            etherTokenAddress = etherToken.address;
+        });
+        afterEach(() => {
+            zeroEx.etherToken.unsubscribeAll();
+        });
+        // Hack: Mocha does not allow a test to be both async and have a `done` callback
+        // Since we need to await the receipt of the event in the `subscribe` callback,
+        // we do need both. A hack is to make the top-level a sync fn w/ a done callback and then
```
Comment (fabioberger) [inline packages/0x.js/test/ether_token_wrapper_test.ts]: `top-level a sync` -> `top-level async`

File: packages/0x.js/src/contract_wrappers/ether_token_wrapper.ts [hunk 2]
```diff
@@ -75,15 +87,76 @@ export class EtherTokenWrapper extends ContractWrapper {
...
+        super.unsubscribeAll();
+    }
     private _invalidateContractInstance(): void {
-        delete this._etherTokenContractIfExists;
+        this.unsubscribeAll();
+        this._etherTokenContractsByAddress = {};
     }
     private async _getEtherTokenContractAsync(etherTokenAddress: string): Promise<EtherTokenContract> {
+        let etherTokenContract = this._etherTokenContractsByAddress[etherTokenAddress];
+        if (!_.isUndefined(etherTokenContract)) {
```
[the remark this answers is not in the document]
Comment (LogvinovLeon) [reply packages/0x.js/src/contract_wrappers/ether_token_wrapper.ts]: For the map to be undefined? That's impossible.
For the lookup result to be undefined? Yes. Lookup returns `T|undefined`
But here we treat it like a cache and return early if found

File: packages/0x.js/test/ether_token_wrapper_test.ts [hunk 3]
```diff
@@ -113,4 +135,205 @@ describe('EtherTokenWrapper', () => {
...
+                    done();
+                };
+                zeroEx.etherToken.subscribe(
+                    etherTokenAddress, EtherTokenEvents.Deposit, indexFilterValues, callback);
+                await zeroEx.etherToken.depositAsync(
+                    etherTokenAddress, depositAmount, addressWithETH,
+                );
+            })().catch(done);
+        });
+        it('Should receive the Withdrawal event when ether is being withdrewn', (done: DoneCallback) => {
```
Comment (fabioberger) [inline packages/0x.js/test/ether_token_wrapper_test.ts]: Typo: `withdrewn` -> `withdrawn`

File: packages/0x.js/test/ether_token_wrapper_test.ts [hunk 4]
```diff
@@ -113,4 +135,205 @@ describe('EtherTokenWrapper', () => {
...
+        });
+        // Hack: Mocha does not allow a test to be both async and have a `done` callback
+        // Since we need to await the receipt of the event in the `subscribe` callback,
+        // we do need both. A hack is to make the top-level async fn w/ a done callback and then
+        // wrap the rest of the test in an async block
+        // Source: https://github.com/mochajs/mocha/issues/2407
+        it('Should receive the Transfer event when tokens are transfered', (done: DoneCallback) => {
+            (async () => {
+                const callback = (err: Error, logEvent: DecodedLogEvent<TransferContractEventArgs>) => {
+                    expect(logEvent).to.not.be.undefined();
```
Comment (fabioberger) [inline packages/0x.js/test/ether_token_wrapper_test.ts]: For all these tests can we also assert that `err` is `null`?

File: packages/0x.js/test/ether_token_wrapper_test.ts [hunk 5]
```diff
@@ -113,4 +135,205 @@ describe('EtherTokenWrapper', () => {
...
+                    etherTokenAddress, depositAmount, addressWithETH,
+                );
+                zeroEx.etherToken.subscribe(
+                    etherTokenAddress, EtherTokenEvents.Withdrawal, indexFilterValues, callback);
+                await zeroEx.etherToken.withdrawAsync(
+                    etherTokenAddress, withdrawalAmount, addressWithETH,
+                );
+            })().catch(done);
+        });
+        it('Outstanding subscriptions are cancelled when zeroEx.setProvider called', (done: DoneCallback) => {
```
Comment (fabioberger) [inline packages/0x.js/test/ether_token_wrapper_test.ts]: `it('should cancel outstanding subscriptions when ZeroEx.setProvider is called')`

File: packages/0x.js/test/utils/token_utils.ts [hunk 6]
```diff
@@ -3,7 +3,7 @@ import * as _ from 'lodash';
 import {InternalZeroExError, Token} from '../../src/types';
 
 const PROTOCOL_TOKEN_SYMBOL = 'ZRX';
-const WETH_TOKEN_SYMBOL = 'WETH';
+const WETH_TOKEN_SYMBOL = 'WETH9';
```
Comment (fabioberger) [inline packages/0x.js/test/utils/token_utils.ts]: According to the Solidity source code, the symbol should still be `WETH`... What is going on here?

File: packages/contracts/deploy/migrations/migrate.ts [hunk 7]
```diff
@@ -68,6 +69,18 @@ export const migrator = {
                 gas: addTokenGasEstimate,
             },
         );
+        await tokenReg.addToken.sendTransactionAsync(
+            makerEtherToken.address,
+            'Maker Ether Token',
+            'WETH9',
```
Comment (fabioberger) [inline packages/contracts/deploy/migrations/migrate.ts]: This should be `WETH`

File: packages/contracts/deploy/migrations/migrate.ts [hunk 8]
```diff
@@ -68,6 +69,18 @@ export const migrator = {
                 gas: addTokenGasEstimate,
             },
         );
+        await tokenReg.addToken.sendTransactionAsync(
+            makerEtherToken.address,
+            'Maker Ether Token',
```
Comment (fabioberger) [inline packages/contracts/deploy/migrations/migrate.ts]: According to the smart contract, this should be `Wrapped Ether`

File: packages/0x.js/test/ether_token_wrapper_test.ts [hunk 9]
```diff
@@ -228,7 +232,7 @@ describe('EtherTokenWrapper', () => {
                 );
             })().catch(done);
         });
-        it('Outstanding subscriptions are cancelled when zeroEx.setProvider called', (done: DoneCallback) => {
+        it('should cancel outstanding subscriptions when ZeroEx.setProvider is called', (done: DoneCallback) => {
```
Comment (fabioberger) [inline packages/0x.js/test/ether_token_wrapper_test.ts]: My bad, this should be `zeroEx.setProvider`

File: packages/contracts/util/artifacts.ts [hunk 10]
```diff
@@ -20,8 +19,7 @@ export class Artifacts {
         this.ZRXToken = artifacts.require('ZRXToken');
         this.DummyToken = artifacts.require('DummyToken');
         this.DummyTokenV2 = artifacts.require('DummyToken_v2');
-        this.EtherToken = artifacts.require('EtherToken');
-        this.EtherTokenV2 = artifacts.require('EtherToken_v2');
+        this.EtherToken = artifacts.require('WETH9');
```
Comment (fabioberger) [inline packages/contracts/util/artifacts.ts]: Are there no tests associated with `EtherToken_v2`?

**Pull request merged**